distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/checks/EntitlementChecksService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.runtime.checks;
+
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.checks.EntitlementProvider;
+import org.elasticsearch.entitlement.runtime.api.Entitlement;
+import org.elasticsearch.entitlement.runtime.api.FlagEntitlement;
+import org.elasticsearch.entitlement.runtime.api.NotEntitledException;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Collections.newSetFromMap;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isActive;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isFrozen;
+
+/**
+ * Implementation of the {@link EntitlementChecks} interface.
+ * The trampoline module loads this object via SPI.
+ */
+public class EntitlementChecksService implements EntitlementChecks {
+    /**
+     * @return the same instance of {@link EntitlementChecksService} returned by {@link EntitlementProvider}.
+     */
+    public static EntitlementChecksService get() {
+        return (EntitlementChecksService) EntitlementProvider.checks();
+    }
+
+    private final Set<Module> entitledToExit = newSetFromMap(new ConcurrentHashMap<>());
+
+    public void grant(Module module, Entitlement e) {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        switch (e) {
+            case FlagEntitlement f -> {
+                switch (f) {
+                    case EXIT_JVM -> entitledToExit.add(module);
+                }
+            }
+        }
+    }
+
+    /**
+     * Mainly for testing purposes.
+     */
+    public void revokeAll() {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        entitledToExit.clear();
+    }
+
+    /**
+     * Causes entitlements to be enforced.
+     */
+    public void activate() {
+        isActive = true;
+    }
+
+    /**
+     * Disallows changes to the entitlement grants.
+     */
+    public void freeze() {
+        isFrozen = true;
+    }
+
+    @Override
+    public void checkSystemExit(Class<?> callerClass, System system, int status) {
```

Comment:
I find it quite limiting that we need to define a new method for all the places (methods) we want to instrument.
Again, this probably does not matter for now (1 method), but I would just leave out this for now and wire the methods to instrument directly to the "generic" check (checkFlagEntitlement) in the Transformer/Instrumenter.

Reply:
We need to list all the methods somewhere.

Reply:
Eventually yes, for now we have 1 method. Choosing a way to do that now, and a very opinionated one, is a bit too binding. Especially if we find a way to partially automate the process of identifying the places to instrument, having then to translate each of them to an interface method + implementation is going to be cumbersome. 
Again, it could be a solution, but I think it's best not to have this now in this PR.

Reply:
As we discussed during the sync, I'd have done in another way (likely, the way you originally tried), to link the instrumentation-site definition and the check methods in a different way (e.g. a static class linking functions to instrument to check functions, and keep those in the interface, and the interface compact).
Here the "link" is provided in the implementation. It's OK, especially for a single method here -- it would be easy to change.
I'm going to resolve this and the other linked comments